cmd/forwarder/auth_test.go

```diff
@@ -0,0 +1,208 @@
+package main
+
+import (
+	"encoding/json"
+	"net/http"
+	"testing"
+	"time"
+
+	"github.com/elliotchance/redismock"
+	"github.com/go-redis/redis"
+	metrics "github.com/rcrowley/go-metrics"
+	"github.com/stretchr/testify/assert"
+)
+
+func noSecretsRedis() redis.Cmdable {
+	r := redismock.NewMock()
+	r.On("HGetAll").Return(redis.NewStringStringMapCmd("HGetAll"))
+	return r
+}
+
+func oneSecretRedis() redis.Cmdable {
+	r := redismock.NewMock()
+	m := make(map[string]string)
+	creds := make([]Credential, 0, 1)
+	creds = append(creds, Credential{Stage: "current", Hmac: hmacEncode("hmacKey", "newpassword")})
+	m["newuser"] = marshal(creds)
+	cmd := redis.NewStringStringMapResult(m, nil)
+	r.On("HGetAll").Return(cmd)
+	return r
+}
+
+func overrideRedis() redis.Cmdable {
+	r := redismock.NewMock()
+	m := make(map[string]string)
+	creds := make([]Credential, 0, 1)
+	creds = append(creds, Credential{Stage: "current", Hmac: hmacEncode("hmacKey", "newpassword")})
+	m["user"] = marshal(creds)
+	cmd := redis.NewStringStringMapResult(m, nil)
+	r.On("HGetAll").Return(cmd)
+	return r
+}
+
+func marshal(creds []Credential) string {
+	b, err := json.Marshal(creds)
+	if err != nil {
+		panic(err)
+	}
+	return string(b)
+}
+
+func defaultCreds() *BasicAuth {
+	ba := NewBasicAuth(metrics.NewRegistry(), "hmacKey")
+	ba.AddPrincipal("user", hmacEncode("hmacKey", "password"), "env")
+	return ba
+}
+
+func newSecretCreds() *BasicAuth {
+	ba := defaultCreds()
+	ba.AddPrincipal("newuser", hmacEncode("hmacKey", "newpassword"), "current")
+	return ba
+}
+
+func overrideCreds() *BasicAuth {
+	ba := NewBasicAuth(metrics.NewRegistry(), "hmacKey")
+	ba.AddPrincipal("user", hmacEncode("hmacKey", "newpassword"), "current")
+	return ba
+}
+
+func TestRefreshAuth(t *testing.T) {
+	tests := map[string]struct {
+		auth          *BasicAuth
+		client        redis.Cmdable
+		config        string
+		expectChanged bool
+		expectError   error
+		expectedCreds *BasicAuth
+	}{
+		"no secrets in passwordmanager": {
+			auth:          defaultCreds(),
+			client:        noSecretsRedis(),
+			config:        "user:password",
+			expectedCreds: defaultCreds(),
+			expectChanged: false,
+		},
+		"new secret added": {
+			auth:          defaultCreds(),
+			client:        oneSecretRedis(),
+			config:        "user:password",
+			expectedCreds: newSecretCreds(),
+			expectChanged: true,
+		},
+		"env creds can be overridden": {
+			auth:          defaultCreds(),
+			client:        overrideRedis(),
+			config:        "user:password",
+			expectedCreds: overrideCreds(),
+			expectChanged: true,
+		},
+	}
+
+	for name, test := range tests {
```

Comment:
Can you not use t.Run() here?

Reply:
@freeformz I found this guide to writing table-based tests in go: https://medium.com/agrea-technogies/basic-testing-patterns-in-go-d8501e360197

It didn't mention `t.Run()`, unlike the docs here: https://golang.org/pkg/testing/

So - yes, I can use `t.Run()` and I'll change the code to do so.  I followed a guide that made no mention of this.

Reply:
I think that guide predated t.Run()